Fe#21/create edit profile UI

jedchernandez opened the pull request:
### What this PR does (required):
- Create edit profile UI
- Update profile model and interface
- Update styling to try and replicate mock up
- Leave TODO comments to address in the future
### Screenshots / Videos (required):
- Owner Edit Profile A:
![Owner Edit Profile UI - A](https://user-images.githubusercontent.com/62577683/130123530-6a8944c6-cbbe-413c-b639-1311efb0efa9.jpg)
- Owner Edit Profile B:
![Owner Edit Profile UI - B](https://user-images.githubusercontent.com/62577683/130123568-5b752ab7-fe8d-4a67-bafc-2b85e77ac3ca.jpg)
- Sitter Edit Profile A:
![Sitter Edit Profile UI - A](https://user-images.githubusercontent.com/62577683/130123619-69f92a95-521e-4108-a6fd-d3e4b4944138.jpg)
- Sitter Edit Profile B:
![Sitter Edit Profile UI - B](https://user-images.githubusercontent.com/62577683/130123646-88d65d9c-d167-4d37-bebd-782523ceebf1.jpg)
### Any information needed to test this feature (required):
- npm run dev to start the server
- npm start to run the app to your localhost
- login using given credentials
- To see either the Owner or Sitter edit profile, it was done programmatically by modifying some variables:
Under Options.tsx, you can modify testType = 'Sitter' or 'Owner': 
![image](https://user-images.githubusercontent.com/62577683/130124212-10bb6cac-680a-4e62-b601-52ba6f6e3547.png)
Under EditProfileForm.tsx, you can modify the Formik initialValues for type: ProfileType.Sitter or ProfileType.Owner:
![image](https://user-images.githubusercontent.com/62577683/130124550-9707e7fb-b898-4c94-8726-02f88833539d.png)
These will not be necessary once integrated.
### Any issues with the current functionality (optional):
- There's a button ui issue for the I'm Available button
- When not toggled:
![NotToggled](https://user-images.githubusercontent.com/62577683/130124797-7310ea0f-ecef-4216-9fda-b95c84000264.jpg)
- When toggled:
![Toggled](https://user-images.githubusercontent.com/62577683/130124812-7c6f3cd0-1eda-48de-8bf3-512d7d3cbabf.jpg)

jedchernandez committed:
feat: Modify availability for profile schema

jedchernandez committed:
feat: Add profile interface

jedchernandez committed:
feat: Add all necessary input

 - CSS still needs to be updated to match mockups

jedchernandez committed:
feat: Create edit profile UI

 - Update profile model and interface
 - Update styling to try and replicate mock up
 - Leave TODO comments to address in the future

jedchernandez committed:
feat: Re-style switch button

 - Fix styling issue
 - Create a custom switch button

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/EditProfile.tsx:
```diff
@@ -1,9 +1,38 @@
+    }: Profile,
+    { setSubmitting }: FormikHelpers<Profile>,
+  ) => {
+    // TODO:
+    // Make an API Call to update profile
```
This is nice!

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/EditProfile.tsx:
```diff
@@ -1,9 +1,38 @@
+  };
+  return (
+    <Grid container>
+      <CssBaseline />
```
This is a [global reset](https://material-ui.com/components/css-baseline/) that is applied to the whole project. Therefore it's best if we move it to the top-most-level component (e.g. `App`) if we want to use it - using it once should be enough.

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/EditProfile.tsx:
```diff
@@ -1,9 +1,38 @@
+  return (
+    <Grid container>
+      <CssBaseline />
+      <Box className={classes.box}>
```
You can apply most styles to a `Box` directly without needing additional `className`s (see [docs](https://material-ui.com/system/basics/#all-inclusive) for more info).

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/EditProfileForm/EditProfileForm.tsx:
```diff
@@ -0,0 +1,353 @@
+      value: 'Other',
+      label: 'Other',
+    },
+  ];
```
These could be moved outside of the component - or even to a separate file.

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/EditProfileForm/EditProfileForm.tsx:
```diff
@@ -0,0 +1,353 @@
+          .max(200, 'The maximum you can charge is $200/hour')
+          .min(1, 'The minimum you can charge is $1/hour')
+          .required('Required'),
+      })}
```
This is really good! In general, I recommend defining these objects before passing them as props to increase readability - there are some concerns about performance as well which are not that relevant to us at the moment.

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/EditProfileForm/EditProfileForm.tsx:
```diff
@@ -0,0 +1,353 @@
+      }}
+      validationSchema={Yup.object().shape({
+        firstName: Yup.string().max(15, 'First name must be 15 or less characters').required('Required'),
+        lastName: Yup.string().max(20, 'Last name must be 20 characters or less').required('Required'),
```
I've seen this more than once - is the character limit defined as a spec? If not, I think limiting characters on name/last name fields aren't always a good idea (or at least we can be a bit more generous with them). Just leaving this as a note, no changes are necessary.

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/EditProfileForm/EditProfileForm.tsx:
```diff
@@ -0,0 +1,353 @@
+                id="dateofBirth"
+                inputVariant="outlined"
+                format="MMMM/dd/yyyy"
+                value={dateOfBirth}
+                onChange={handleBirthday}
```
Is there a reason we don't use `Formik`'s own logic here (e.g. `values.dateOfBirth` and `handleChange`)?

rajivtitus replied:
@jedchernandez

jedchernandez replied:
Updated it to use formik instead!

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/EditProfileForm/EditProfileForm.tsx:
```diff
@@ -0,0 +1,353 @@
+    >
+      {({ handleSubmit, handleChange, values, touched, errors, isSubmitting }) => (
+        <form onSubmit={handleSubmit} className={classes.form} noValidate>
+          <Grid className={classes.gridItemB}>
```
I don't think `Grid` is the right choice of component here since there is no `container`/`item` hierarchy - as is it just acts as a `flex-box` but with potentially unexpected behaviour. We can use `Box`s instead or create a proper `Grid` layout (see [here](https://material-ui.com/components/grid/) for more info). This applies to the rest of the `Grid`s below too.

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/EditProfileForm/EditProfileForm.tsx:
```diff
@@ -0,0 +1,353 @@
+      {({ handleSubmit, handleChange, values, touched, errors, isSubmitting }) => (
+        <form onSubmit={handleSubmit} className={classes.form} noValidate>
+          <Grid className={classes.gridItemB}>
+            <Typography>
+              <h1 className={classes.heading}>Edit Profile</h1>
+            </Typography>
```
We should use `component`/`variant` props on `Typography` for this (you can check out [these examples](https://material-ui.com/components/typography/#typography)).

rajivtitus replied:
@jedchernandez

jedchernandez replied:
I've updated it so that it uses variant instead.

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/EditProfileForm/EditProfileForm.tsx:
```diff
@@ -0,0 +1,353 @@
+          {values.type === ProfileType.Sitter && ( //TODO: Fix the button styling
+            <Grid className={`${classes.gridItemA} ${classes.toggle}`}>
+              <label className={classes.label}>{`I'M AVAILABLE`}</label>
+              <AvailableSwitch id="I'M AVAILABLE" checked={isAvailable} onChange={handleToggle} name="I'M AVAILABLE" />
```
I'd recommend keeping a consistent casing (e.g. `camelCase`) when assigning `id`s to elements.

rajivtitus replied:
@jedchernandez

jedchernandez replied:
Thanks for pointing that out!

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/EditProfileForm/AvailableSwitch/AvailableSwitch.tsx:
```diff
@@ -0,0 +1,67 @@
+  classes: Styles;
+}
+
+const AvailableSwitch = withStyles((theme: Theme) =>
```
Really nice to see you diving into advanced customization of `Switch` component! Let's make sure we don't spend too much time on it though as customizing Material UI components can be a pain from time to time :D.

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/EditProfileForm/AvailableSwitch/AvailableSwitch.tsx:
```diff
@@ -0,0 +1,67 @@
+    switchBase: {
+      padding: 3,
+      '&$checked': {
+        transform: 'translateX(20px)',
```
This might be why you see two dots when the `Switch` is not checked.

jedchernandez replied:
I found the error somewhere else. I unnecessarily using a styled wrapper on the component. Once I removed it, it was working.

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/EditProfileForm/EditProfileForm.tsx:
```diff
@@ -0,0 +1,353 @@
+                id="availableHoursPerWeek"
+                select
+                margin="normal"
+                InputLabelProps={{
+                  shrink: true,
+                }}
```
Since you don't provide a `label` to the `TextField` prop, I would imagine this wouldn't actually be needed - is there a caveat or similar that this is needed for or can we remove it?

rajivtitus replied:
@jedchernandez

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/EditProfileForm/EditProfileForm.tsx:
```diff
@@ -0,0 +1,353 @@
+          </Grid>
+          {values.type === ProfileType.Sitter && ( //TODO: Fix the button styling
+            <Grid className={`${classes.gridItemA} ${classes.toggle}`}>
+              <label className={classes.label}>{`I'M AVAILABLE`}</label>
```
We can optionally pass `htmlFor` props to these labels, [this stackoverflow post](https://stackoverflow.com/a/18432439) explains what that does briefly.

jedchernandez replied:
I forgot to change the label tags to typography.

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/EditProfileForm/EditProfileForm.tsx:
```diff
@@ -0,0 +1,353 @@
+                  shrink: true,
+                }}
+                name="availableHoursPerWeek"
+                helperText={touched.availableHoursPerWeek ? errors.availableHoursPerWeek : ''}
```
I think using the `&&` syntax for these would be better?

jedchernandez replied:
Right! Updated it!

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/EditProfileForm/EditProfileForm.tsx:
```diff
@@ -0,0 +1,353 @@
+const EditProfileForm = ({ handleSubmit }: Props): JSX.Element => {
+  const classes = useStyles();
+
+  const [isAvailable, setIsAvailable] = useState(true);
+  const handleToggle = () => {
+    isAvailable ? setIsAvailable(false) : setIsAvailable(true);
+  };
```
Do we need this state, is there a problem using `Formik`'s state for these instead?

rajivtitus replied:
@jedchernandez

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/EditProfileForm/EditProfileForm.tsx:
```diff
@@ -0,0 +1,353 @@
+    },
+  ];
+
+  const [dateOfBirth, setDateOfBirth] = useState(new Date());
+  const handleBirthday = (date: any) => {
+    setDateOfBirth(date);
+  };
```
Similarly, is this needed?

rajivtitus replied:
@jedchernandez

jedchernandez replied:
I've updated it to use formik instead.

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/EditProfileForm/useStyles.ts:
```diff
@@ -0,0 +1,64 @@
+    marginBottom: '1%',
+  },
+  gridItemA: {
+    display: 'flex',
+    flexFlow: 'row wrap',
+    justifyContent: 'flex-start',
+    alignContent: 'center',
+    alignItems: 'center',
+    fontWeight: 'bold',
```
You might not need most of these as they are defined as props you can pass to the `Grid` component (listed [here](https://material-ui.com/api/grid/#props)).

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/EditProfileForm/useStyles.ts:
```diff
@@ -0,0 +1,64 @@
+    fontWeight: 'bold',
+  },
+  gridItemB: {
+    display: 'flex',
+    flexDirection: 'row',
+    alignItems: 'center',
+    justifyContent: 'center',
+    margin: '5% 0 2% 0',
+    fontWeight: 'bold',
```
Similarly these too.

ugur-akin commented on client/src/components/ProfileSettings/EditProfile/useStyles.ts:
```diff
@@ -1,8 +1,11 @@
 
-const useStyles = makeStyles(() => ({}));
+const useStyles = makeStyles(() => ({
+  box: {
+    backgroundColor: 'white',
+    height: '100%',
+    width: '100%',
+  },
```
Similarly, `Box` components can receive most of the styles applied as props.

ugur-akin commented on client/src/components/ProfileSettings/Options/Options.tsx:
```diff
@@ -19,22 +30,66 @@ const Options = (): JSX.Element => {
+        <ListItem
+          button
+          disableGutters
+          selected={selectedIndex === 0}
+          onClick={(event: React.MouseEvent<HTMLElement>) => handleListItemClick(event, 0)}
```
In my opinion, a better way of doing this would be to use the [`useLocation` hook from `react-router-dom`](https://reactrouter.com/web/api/Hooks/uselocation) and check if the [current pathname](https://reactrouter.com/web/api/location) matches the route.

jedchernandez replied:
The reason why I did this was to use the specific property: **_selected_** for the ListItem so that when you click an option it will be highlighted. However, I did use useLocation to switch between Options to render its own component accordingly. 
Highlighted Option: 
![image](https://user-images.githubusercontent.com/62577683/130504521-9ab14de0-6888-4afa-a755-2d21c92c181d.png)

ugur-akin replied:
I was thinking similarly, you could use 
```
const {pathname} = useLocation();
...
    selected = { pathname === 'route-to-match' }
```
Entirely up to you though, both work 👍🏼

jedchernandez replied:
Ah I see. I keep it as is for now. Thanks!

ugur-akin commented on client/src/components/ProfileSettings/Options/Options.tsx:
```diff
@@ -19,22 +30,66 @@ const Options = (): JSX.Element => {
-          <ListItemText primary="Availability" className={classes.listItem} />
-        </ListItem>
-        <ListItem button disableGutters component={Link} to="/profile/options/payment">
+        {testType === ProfileType.Sitter && ( // TODO: This condition will be modified accordingly once integration is applied
```
I really like these comments - as the project grows and the structure gets more and more complicated, these will come in clutch!

ugur-akin commented on client/src/components/ProfileSettings/Options/Options.tsx:
```diff
@@ -19,22 +30,66 @@ const Options = (): JSX.Element => {
     >
       <List>
-        <ListItem button disableGutters component={Link} to="/profile/options/edit-profile">
+        <ListItem
+          button
+          disableGutters
+          selected={selectedIndex === 0}
+          onClick={(event: React.MouseEvent<HTMLElement>) => handleListItemClick(event, 0)}
+          component={Link}
+          to="/profile/options/edit-profile"
+        >
```
Nice use of `react-router-dom` with `Material-UI`!

ugur-akin reviewed (commented):
Wow, lots of changes here - well done! I left some comments here and there. One more thing about PRs in general, going forward it would be really helpful in terms of project management and code reviewing if we: 
- name our PRs closer to plain english (e.g. `Implementation of Edit Profile UI`),
- [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#manually-linking-a-pull-request-to-an-issue) it's related to,
- and add any additional labels that fit the PR.

rajivtitus commented on client/src/components/ProfileSettings/EditProfile/EditProfileForm/EditProfileForm.tsx:
```diff
@@ -0,0 +1,353 @@
+          )}
+          {values.type === ProfileType.Sitter && (
+            <Grid className={classes.gridItemA}>
+              <label className={classes.label}>AVAILABILITY</label>
```
Reserve usage of capital case letters for abbreviations, you can change the text's case using CSS classes here instead. (Ditto throughout). Please see below for a reference on why to avoid capital casing:

Ref: https://medium.com/@mandy.michael/why-you-shouldnt-write-your-content-in-uppercase-instead-use-css-b03ac2c65b99

jedchernandez committed:
feat: Address comments

jedchernandez commented:
> Wow, lots of changes here - well done! I left some comments here and there. One more thing about PRs in general, going forward it would be really helpful in terms of project management and code reviewing if we:
> 
> * name our PRs closer to plain english (e.g. `Implementation of Edit Profile UI`),
> * [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#manually-linking-a-pull-request-to-an-issue) it's related to,
> * and add any additional labels that fit the PR.

Thanks. There weren't any contribution guidelines before. I'll follow this going forward.

jedchernandez committed:
fix: Push a missed commit for restyling

rajivtitus reviewed (commented):
Hey @jedchernandez , good work making requested changes. However, you may have missed a couple of comments, please have a look and address. Thanks

jedchernandez committed:
refactor: Address missed comments

jedchernandez commented:
@rajivtitus Thanks. Did some refactoring and updated it. The other comments that aren't labeled **_Outdated_** are just descriptions or were already addressed; 
I.e.
![image](https://user-images.githubusercontent.com/62577683/130655832-8df9c566-4889-4c09-9db4-4cd37947f5e4.png)

rajivtitus commented:
> @rajivtitus Thanks. Did some refactoring and updated it. The other comments that aren't labeled **_Outdated_** are just descriptions or were already addressed;
> I.e.
> ![image](https://user-images.githubusercontent.com/62577683/130655832-8df9c566-4889-4c09-9db4-4cd37947f5e4.png)

I've tagged you in a few comments which still appear unresolved/haven't been replied based on what I see on my end, if you've already addressed them (and it's a github issue) please let me know :)

jedchernandez committed:
feat: Address missed comments

 - Replaced the usage of useState to formik's logic
 - Replaced labels to use typography
 - Fix helperText's logic for formik
 - Modified isAvailable model property type from boolean to string

jedchernandez committed:
fix: Remove input label props

jedchernandez reviewed (commented):
@rajivtitus 
I definitely missed those! I apologize. I've addressed them now. Thanks!
